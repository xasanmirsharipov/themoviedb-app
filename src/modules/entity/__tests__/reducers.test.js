import React from 'react';
import Actions from '../actions';
import Reducer from '../reducers';

describe('Entity Reducers', () => {

	const entity = 'movieEntity';
	const entityType = 'movieEntityType';

	describe('LoadALL with empty data', () => {

		const state = {};

		it('Request', () => {

			const payload = {
				entity,
				name: entityType
			};

			expect(Reducer(state, Actions.LoadAll.request(payload))).toEqual({
				[entity]: {
					[entityType]: {
						isFetched: false
					}
				}
			});
		});

		it('Success', () => {

			const payload = {
				entity,
				name: entityType,
				ids: [1, 2, 3],
				meta: {
					totalCount: 0,
					pageCount: 1,
					currentPage: 1,
					perPage: 20
				}
			};

			const reducer = Reducer(state, Actions.LoadAll.success(payload));

			expect(reducer).toEqual({
				[entity]: {
					[payload.name]: {
						isFetched: true,
						ids: payload.ids,
						meta: payload.meta
					}
				}
			})
		});

		it('Failure', () => {

			const payload = {
				entity,
				name: entityType,
				error: 'error text'
			};

			const reducer = Reducer(state, Actions.LoadAll.failure(payload));

			expect(reducer).toEqual({
				[entity]: {
					[payload.name]: {
						isFetched: true,
						error: payload.error
					}
				}
			})
		});

	});

	describe('LoadALL with initial state', () => {

		const state = {
			[entity]: {
				[entityType]: {
					isFetched: true,
					ids: [1, 2, 3],
					meta: {
						totalCount: 0,
						pageCount: 1,
						currentPage: 1,
						perPage: 20
					}
				}
			}
		};

		it('REQUEST', () => {

			const payload = {
				entity,
				name: entityType
			};

			const reducer = Reducer(state, Actions.LoadAll.request(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[payload.name]: {
						...state[payload.entity][payload.name],
						isFetched: false
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

		it('SUCCESS', () => {

			const payload = {
				entity,
				name: entityType,
				ids: [4, 5, 6],
				meta: {
					totalCount: 0,
					pageCount: 1,
					currentPage: 1,
					perPage: 20
				}
			};

			const reducer = Reducer(state, Actions.LoadAll.success(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[payload.name]: {
						...state[payload.entity][payload.name],
						isFetched: true,
						ids: [
							...state[payload.entity][payload.name].ids,
							...payload.ids
						],
						meta: payload.meta
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

		it('FAILURE', () => {

			const payload = {
				entity,
				name: entityType,
				error: 'error text'
			};

			const reducer = Reducer(state, Actions.LoadAll.failure(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[payload.name]: {
						...state[payload.entity][payload.name],
						isFetched: true,
						error: payload.error
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

	});

	describe('LoadOne with empty data', () => {

		const state = {};

		const id = 1;

		it('Request', () => {

			const payload = {
				entity,
				name: entityType,
				id
			};

			expect(Reducer(state, Actions.LoadOne.request(payload))).toEqual({
				[entity]: {
					[`${payload.name}One`]: {
						isFetched: false
					}
				}
			});
		});

		it('Success', () => {

			const payload = {
				entity,
				name: entityType,
				id
			};

			const reducer = Reducer(state, Actions.LoadOne.success(payload));

			expect(reducer).toEqual({
				[entity]: {
					[`${payload.name}One`]: {
						isFetched: true,
						id: payload.id
					}
				}
			})
		});

		it('Failure', () => {

			const payload = {
				entity,
				name: entityType,
				error: 'error text'
			};

			const reducer = Reducer(state, Actions.LoadOne.failure(payload));

			expect(reducer).toEqual({
				[entity]: {
					[`${payload.name}One`]: {
						isFetched: true,
						error: payload.error
					}
				}
			})
		});

	});

	describe('LoadOne with initial state', () => {

		const state = {
			[entity]: {
				[`${entityType}One`]: {
					isFetched: true,
					id: 1,
					meta: {
						totalCount: 0,
						pageCount: 1,
						currentPage: 1,
						perPage: 20
					}
				}
			}
		};

		const id = 1;

		it('REQUEST', () => {

			const payload = {
				entity,
				name: entityType,
				id
			};

			const reducer = Reducer(state, Actions.LoadOne.request(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[`${payload.name}One`]: {
						...state[payload.entity][`${payload.name}One`],
						isFetched: false
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

		it('SUCCESS', () => {

			const payload = {
				entity,
				name: entityType,
				id: 1,
				meta: {
					totalCount: 0,
					pageCount: 1,
					currentPage: 1,
					perPage: 20
				}
			};

			const reducer = Reducer(state, Actions.LoadOne.success(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[`${payload.name}One`]: {
						...state[payload.entity][`${payload.name}One`],
						isFetched: true,
						id: payload.id,
						meta: payload.meta
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

		it('FAILURE', () => {

			const payload = {
				entity,
				name: entityType,
				error: 'error text'
			};

			const reducer = Reducer(state, Actions.LoadOne.failure(payload));

			expect(reducer).toEqual({
				[payload.entity]: {
					...state[payload.entity],
					[`${payload.name}One`]: {
						...state[payload.entity][`${payload.name}One`],
						isFetched: true,
						error: payload.error
					}
				}
			});

			expect(reducer).toMatchSnapshot();
		});

	});

});