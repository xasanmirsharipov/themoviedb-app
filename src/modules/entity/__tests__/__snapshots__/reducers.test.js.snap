// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Reducers LoadALL with initial state FAILURE 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityType": Object {
      "error": "error text",
      "ids": Array [
        1,
        2,
        3,
      ],
      "isFetched": true,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadALL with initial state REQUEST 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityType": Object {
      "ids": Array [
        1,
        2,
        3,
      ],
      "isFetched": false,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadALL with initial state REQUEST 2`] = `
Object {
  "movieEntity": Object {
    "movieEntityTypeOne": Object {
      "ids": Array [
        1,
        2,
        3,
      ],
      "isFetched": false,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadALL with initial state SUCCESS 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityType": Object {
      "ids": Array [
        1,
        2,
        3,
        4,
        5,
        6,
      ],
      "isFetched": true,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadOne with initial state FAILURE 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityTypeOne": Object {
      "error": "error text",
      "id": 1,
      "isFetched": true,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadOne with initial state REQUEST 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityTypeOne": Object {
      "id": 1,
      "isFetched": false,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;

exports[`Entity Reducers LoadOne with initial state SUCCESS 1`] = `
Object {
  "movieEntity": Object {
    "movieEntityTypeOne": Object {
      "id": 1,
      "isFetched": true,
      "meta": Object {
        "currentPage": 1,
        "pageCount": 1,
        "perPage": 20,
        "totalCount": 0,
      },
    },
  },
}
`;
